rfc6979/src/lib.rs

```diff
@@ -0,0 +1,87 @@
+//! Support for computing deterministic DSA or ECDSA ephemeral scalar (`k`).
+//!
+//! Implementation of the algorithm described in RFC 6979 (Section 3.2):
+//! <https://tools.ietf.org/html/rfc6979#section-3>
+//!
+//! The generics `I`, `O`, and `T` are merely so checks can be optimized away if all are
+//! fixed-size arrays.
+
+pub use hmac;
+
+use hmac::digest::{BlockInput, FixedOutput, Reset, Update};
+
+pub mod hmac_drbg;
+
+pub struct KGenerator<D, I>
+where
+    D: BlockInput + FixedOutput + Clone + Default + Reset + Update,
+    I: AsRef<[u8]>,
+{
+    modulus: I,
+    rng: hmac_drbg::HmacDrbg<D>,
+}
+
+impl<D, I> KGenerator<D, I>
+where
+    D: BlockInput + FixedOutput + Clone + Default + Reset + Update,
+    I: AsRef<[u8]>,
+{
+    /// The `secret` is the secret part of the DSA private key. `digest` is the hash of the
+    /// message being signed, in general using the same hash algorithm as the one passed to
+    /// `KGenerator` as a generic parameter.
+    ///
+    /// Both should have been truncated to the same number of bits as needed to represent the
+    /// modulus, interpreted as a big-endian number, reduced to being below the modulus by
+    /// subtracting the modulus if it is larger than the modulus, and then encoding as a big
+    /// endian number, padded with leading zero bytes to the same length as the modulus.
+    /// See RFC6979 for details.
+    ///
+    /// The modulus must be in big endian format with no padding (i.e. first byte must not be 0).
+    ///
+    /// This function will panic if these requirements are not met.
+    pub fn new<T: AsRef<[u8]>>(modulus: I, secret: T, digest: T, additional_data: &[u8]) -> Self {
```

Comment:
Separate generics `I` and `T` is a bit ugly here, I think the idea was to allow user to pick an owned or borrowed representation for the modulus?